docs/roadmap/scaling-2019.md

```diff
@@ -0,0 +1,115 @@
+
+However the Autoscaler process is more stateful. It maintains request statistics over a window of time. And it must process data from the Revision Pods continuously to maintain that window of data. It is part of the running system all the time, not just when scaled to zero. As the number of Revisions increase and the number of Pods in each Revision increases, the CPU and memory requirements will exceed that available to a single process. So some sharding is necessary.
+
+**Our goal is to shard Revisions across Autoscaler replicas (e.g. 2x the Replica count means 1/2 the load on each Autoscaler). And for autoscaling to be unaffected by an individual Autoscaler termination.**
+
+* POC: Kenny Leung (Google)
+* Github: [Project 19](https://github.com/knative/serving/projects/19)
+
+### Streaming Autoscaling
+
+In addition to being always available, Web applications are expected to be responsive. Long-lived connections and streaming protocols like Websockets and HTTP2 are essential. Knative Serving accepts HTTP2 connections. And will serve requests multiplexed within the connection. But the autoscaling subsystem doesn't quite know what to do with those connections. It sees each connection as continuous load on the system and so will autoscale accordingly.
+
+But the actual load is in the stream within the connection. So the metrics reported to the Autoscaler should be based on the number of concurrent *streams*. This requires some work in the Queue proxy to crack open the connection and emit stream metrics.
+
+Additionally, concurrency limits should be applied to streams, not connections. So containers which can handle only one request at at time should still be able to serve HTTP2. The Queue proxy will just allow one stream through at a time.
+
+**Our goal is 1) to support HTTP2 end-to-end while scaling on concurrent streams and in this mode 2) enforce concurrency limits on streams (not connections).**
+
+* POC: Markus Thömmes (Red Hat)
+* Github: [Project 16](https://github.com/knative/serving/projects/16)
+
+### Pluggability and HPA
+
+This is work remaining from 2018 to add CPU-based autoscaling to Knative and provide an extension point for further customizing the autoscaling sub-system. Remaining work includes:
+
+1. metrics pipeline relayering to scrape metrics from Pods ([#1927](https://github.com/knative/serving/issues/1927))
+2. adding a `window` annotation to allow for further customization of the KPA autoscaler ([#2909](https://github.com/knative/serving/issues/2909))
+3. implementing scale-to-zero for CPU-scaled workloads ([#3064](https://github.com/knative/serving/issues/3064))
+
+**Our goal is to have a cleanly-layered, extensible autoscaling sub-system which fully supports concurrency and CPU metrics (including scale-to-zero).**
```

Comment:
I'd like to separate out Pluggability from HPA.

1. The autoscaler should be pluggable, which includes being able to write an out-of-tree autoscaler that can be sensitive to (at least) the same request metrics as the KPA.  The HPA can be a proof point of success here, but work on the HPA itself is separate and this work should be successful if another implementation (call it YOLO) comes along and is able to do this and HPA cannot.  Frankly, [like Build](https://github.com/knative/serving/blob/fa1aff18a9b549e79e41cf0b34f66b79c3da06b6/test/controller/main.go#L69), I'd kind of like to see a toy implementation (YOLO!) that we can exercise in our e2e tests to verify this.

2. We have an HPA-based implementation of ^^ that can scale 1->N->1 based on request metrics, but also based on CPU and Memory (what about GPU?).  HPA should have e2e and performance testing.  If pushing on K8s HPA supporting scale-to-zero is a thing, it should be driven by the owner of this.

I think `2.` is a long-term aspiration to unify autoscalers (but not needed for `knative/serving` "v1").  I'm on the fence about `1.`s necessity for "v1", but I think the pieces involved all reflect good architectural hygiene and enable things like `2.` post-"v1", it also enables folks to customize their installs to do exciting things.

Reply:
@mattmoor the primary thing I would like to get out of this line item is HPA scale-to-zero.  We can implement scale-to-zero in the Knative control plane without changes to the K8s HPA itself.  As a side-effect, we will be advancing the pluggability of the autoscaling system.  See https://github.com/knative/serving/issues/3064

Reply:
@josephburnett scaling to zero on CPU doesn't make sense to me (div0), so I'll have to read through the issue you linked.  thanks for the pointer.

Reply:
@mattmoor I've broken this into two separate line items: 1) pluggability and 2) hpa integration.  Good point!